### PR TITLE
Remove repetitive code in module_sf_urban.F

### PR DESCRIPTION
**Pull request by epn09**
TYPE: enhancement

KEYWORDS: urban physics

SOURCE: Do Ngoc Khanh (Tokyo Institute of Technology)

DESCRIPTION OF CHANGES:
Problem: There was a code block repeated 12 times.

Solution: Purely refactor, no change in algorithm.

LIST OF MODIFIED FILES:
M       phys/module_sf_urban.F

TESTS CONDUCTED

1. Confirmed bit-by-bit indentical output, before vs after.
2. Jenkins tests are all passing.

RELEASE NOTE: Remove repetitive code in module_sf_urban.F

### CODE REVIEW

File: phys/module_sf_urban.F [hunk 1]
```diff
@@ -2647,341 +2647,42 @@ SUBROUTINE urban_var_init(ISURBAN, TSURFACE0_URB,TLAYER0_URB,TDEEP0_URB,IVGTYP,
...
-                  CONTINUE
-                ELSE
-                  WRITE(mesg,*) 'WARNING, FRC_URB2D = 0 BUT IVGTYP IS URBAN'
-                  WRITE_MESSAGE(mesg)
-                  WRITE(mesg,*) 'WARNING, THE URBAN FRACTION WILL BE READ FROM URBPARM.TBL'
-                  WRITE_MESSAGE(mesg)
-                  FRC_URB2D(I,J) = FRC_URB_TBL(UTYPE_URB)
-                ENDIF
-              SWITCH_URB=1
-            ENDIF
```
Comment (davegill) [inline phys/module_sf_urban.F:2725]: Should this repeated block appear one time? Do you remove all occurrences?

**Comment by davegill** [comment]
@epn09 
It looks as if there is missing code that needs to be included at least one time.

**Comment by dudhia** [comment]
@Cenlin He ***@***.***> This refers to the LCZ changes where each
LCZ has a duplicate code section.
Can you check this?

**Comment by cenlinhe** [comment]
No, these parts should not be removed. These if-statements are used to assign the correct values to UTYPE_URB2D and UTYPE_URB according to the LCZ value in each grid. But I agree that there are some repeated codes within each if-statement that could be further refined. However, since this will not affect the model efficiency or accuracy, I suggest not doing any changes to this part right now, or at least the person who submitted this PR should first check with the original code developer @andreazonato

File: phys/module_sf_urban.F (hunk 1 above)
Comment (epn09) [reply phys/module_sf_urban.F:2725]: @davegill One occurrence is kept in the end. I'll do one test with the urban options, but where should I upload the test result to?

**Comment by epn09** [comment]
@cenlinhe I understand the purpose of the original code but from LCZ_1 to LCZ_11, the difference is only in the first two lines (written below), the following lines are purely copy and paste.
```
           IF( IVGTYP(I,J) == LCZ_1) THEN
               UTYPE_URB2D(I,J) = 1  ! high-intensity
```

File: phys/module_sf_urban.F [hunk 2]
```diff
@@ -2647,341 +2647,42 @@ SUBROUTINE urban_var_init(ISURBAN, TSURFACE0_URB,TLAYER0_URB,TDEEP0_URB,IVGTYP,
...
-!!!!!
-!!!!!
-           IF( IVGTYP(I,J) == LCZ_9) THEN
-               UTYPE_URB2D(I,J) = 9  ! high-intensity
-               UTYPE_URB = UTYPE_URB2D(I,J)  ! high-intensity
+            IF (SWITCH_URB == 1) THEN
```
Comment (epn09) [inline phys/module_sf_urban.F:2684]: @davegill The diff is too large. The "missing code" that you mentioned starts from here.

**Comment by cenlinhe** [comment]
> 

@epn09 Yes, you are right. Could you please do a test to see if your refined code produces exactly the same results as the original code? If so, we can approve the change.

**Comment by epn09** [comment]
@cenlinhe I have split the original commit into 6 smaller commits. Is it easier to understand for you?
I'm also running a test to confirm if the refined version produce the same result.

**Comment by epn09** [comment]
@cenlinhe @davegill I have conducted a test with urban options and confirm that the output of the original code and the modified code are exactly the same (bit-by-bit equal). How should I share the result?

**Comment by andreazonato** [comment]
Hi all,

I think these modifications make sense, they decrease the computational time for sure.

The only doubt I have regards  the first SWITCH_URB=1. Is it in the right position?

Andrea

**Comment by epn09** [comment]
Hi @andreazonato,

For your concern, you can take a look at this comit https://github.com/wrf-model/WRF/pull/1593/commits/a8d55d720c551fe08115b31fa914185131159267 or the explanation below.

Originally, the code was
```fortran
SWITCH_URB = 0

IF (IVGTYP(I,J) == ISURBAN) THEN
  ! do some thing for ISURBAN
  SWITCH_URB = 1
ENDIF
IF (IVGTYP(I,J) == LCZ_1) THEN
  ! do some thing for LCZ_1
  SWITCH_URB = 1
ENDIF

IF (SWITCH_URB == 0) THEN
  CONTINUE
ELSE
  ! do something for non-urban
ENDIF
```

The first two ifs are mutually exclusive, so we can combine with `ELSE`.
```fortran
SWITCH_URB = 0

IF (IVGTYP(I,J) == ISURBAN) THEN
  ! do some thing for ISURBAN
  SWITCH_URB = 1
ELSE IF (IVGTYP(I,J) == LCZ_1) THEN
  ! do some thing for LCZ_1
  SWITCH_URB = 1
ENDIF

IF (SWITCH_URB == 0) THEN
  CONTINUE
ELSE
  ! do something for non-urban
ENDIF
```

The role of `SWITCH_URB` is to check if we've hit any of ISURBAN, LCZ_1, ..., LCZ_11, thus, initially we can set it to 1, and changed it back to 0 if none of the ifs match.
```fortran
SWITCH_URB = 1

IF (IVGTYP(I,J) == ISURBAN) THEN
  ! do some thing for ISURBAN
ELSE IF (IVGTYP(I,J) == LCZ_1) THEN
  ! do some thing for LCZ_1
ELSE
  SWITCH_URB = 0
ENDIF

IF (SWITCH_URB == 0) THEN
  CONTINUE
ELSE
  ! do something for non-urban
ENDIF
```

**Comment by davegill** [comment]
@andreazonato @cenlinhe
Folks,
Have your concerns been addressed? We would like to move these PRs into the develop branch if they are ready. Let us know if you have other issues, or are able to accept these mods, AS-IS. A simple "I am OK with these mods" is sufficient.

**Comment by epn09** [comment]
@davegill @andreazonato @cenlinhe 
Hi, is there any update on this?

**Comment by andreazonato** [comment]
Yes, in my opinion it makes sense. I agree with the modifications!

**Comment by cenlinhe** [comment]
It looks good to me too.

**Comment by weiwangncar** [comment]
@cenlinhe Can you approve this PR? Click on the 'Files changed' option and see if you can see 'Review changes' in a green box on top right. If not, I will do that.

**Review by cenlinhe — APPROVED** [review]
This re-structuring of the code looks good and the developer has also confirmed the same results before and after this re-structure.

**Comment by davegill** [comment]
@epn09 
Would you please add you affiliation to the pull request commit message?
Go to https://github.com/wrf-model/WRF/pull/1593

As shown below, click on the three dots, then select "Edit".
Screen Shot 2021-12-28 at 10 50 50 AM

Once you have added your affiliation, at the bottom right, click on the green "Update comment" button.

Screen Shot 2021-12-28 at 10 52 18 AM

**Comment by cenlinhe** [comment]
@epn09 To keep a record, could you please also send us the results of your bit-by-bit comparison before and after the change? Any summary tables, figures, and/or small-size data files will be good. My email: cenlinhe@ucar.edu

**Review by weiwangncar — APPROVED** [review]
The LSM/urban experts have said ok to this code change.

**Comment by epn09** [comment]
@davegill I have added my affiliation.

**Comment by epn09** [comment]
@davegill I have sent a copy of my test output to @cenlinhe and he said that it's acceptable.

**Review by davegill — APPROVED** [review]

**Pull request merged**